### DISCUSSION TITLE
Strange Ruby behavior in Mongoid 8.0 that doesn't happen in Mongoid 7.5

### DISCUSSION THREAD

**Discussion by rsosag** [mongodb/mongoid#5902] [Q&A] [answered]
I need help understanding the  behavior using Ruby and Mongoid 8.0 with `accepts_nested_attributes_for` and an `attr_accessor` in the `embeds_many`

I have a Customer model (simplified)

```ruby
 class Customer
   include Mongoid::Document
   
   field :name,        type: String
   field :nickname,    type: String
   
   embeds_many :business_settings, class_name: 'BusinessSettings', cascade_callbacks: true
   accepts_nested_attributes_for :business_settings

   delegate(*business_settings_attributes, to: :current_business_settings, allow_nil: true)
   
   def self.business_settings_attributes
     business_settings_attrs = BusinesSettings.attribute_names.flat_map { |a| [a, "#{a}=", "#{a}?", "#{a}_changed?"] }
     customer_attrs = Customer.attribute_names.flat_map { |a| [a, "#{a}=", "#{a}?", "#{a}_changed?"] }

     attrs = (business_settings_attrs -  customer_attrs)
     attrs << 'role_or_custom_role_id'
     attrs << 'role_or_custom_role_id='
     attrs << 'status'
     attrs << 'status='
     attrs.map(&:to_sym)
   end
....
```
and the another model (simplified)

```ruby
 class BusinessSettings
   include Mongoid::Document

   attr_accessor :role_or_custom_role_id
   
   field :role,  type: String
   field :status, type: String
   
   embedded_in :customer

   before_validation :set_role
....
```
The problem is that when I try to update a customer, changing the virtual attribute `role_or_custom_role_id`, something like
```ruby
customer.role_or_custom_role_id = 'new_id'
customer.save
```
the callback `before_validation :set_role` does not execute and the embedded is not updated, but if I change another attribute, not only the `attr_accessor`, then the callback is executed and the embedded is updated

```ruby
customer.role_or_custom_role_id = 'new_id'
customer.status = 'new_status'
customer.save
```
the code above does execute the callback and the embedded is updated

Whit Mongoid 7.5 changing only the virtual attribute `role_or_custom_role_id` works fine, but with Mongoid 8.0 it does not work for me.

**NOTE:** It's like in Mongoid 8.0, it doesn't detect changes in the embedded model when only the `attr_accessor` changes, if I do a `customer.current_business_settings.touch` then it executes the callback and saves the update

Can someone help me understand? 🙏🏼

**Comment by jamis** [comment, marked as answer]
Hello,

I'm sorry this one was so mysterious! It's a consequence of #5787, which fixed a significant performance issue with validations on associations. 

So, you're right, that the issue has to do with `attr_accessor`. Validations are now only run on associated records if the records have changed. Because setting a value via `attr_accessor` does not mark the record as changed (that only happens when persistable fields are modified), the validation (and before_validation) callbacks aren't being invoked. To work around this, would it be feasible for you to do something like:

```ruby
class BusinessSettings
  include Mongoid::Document

  def role_or_custom_role_id=(value)
    # move the logic from `set_role` to here...
  end

  # ...and then remove the `before_validation` callback?
end
```

**Comment by rsosag** [comment]
@jamis thanks a lot, this solution works for me
